lib/AST/Decl.cpp

```diff
@@ -2099,8 +2099,14 @@ bool NominalTypeDecl::derivesProtocolConformance(ProtocolDecl *protocol) const {
     // Hashable conformance.
     case KnownProtocolKind::Equatable:
     case KnownProtocolKind::Hashable:
-      return enumDecl->hasCases()
-          && enumDecl->hasOnlyCasesWithoutAssociatedValues();
+      // FIXME: This is too lenient; we can only promise to provide a witness
+      // for enums where all associated values conform to the protocol.
+      // Unfortunately, figuring this out requires a TypeChecker and resolving
+      // the argument type of each enum element. Instead, we let
+      // DerivedConformance::derive{Equatable,Hashable} perform the full check
+      // and it outputs its own diagnostics if conformance cannot be
+      // synthesized.
+      return enumDecl->hasCases();
```

Comment:
Since this is my first time really diving into the various compiler components, can you elaborate a little more on what you mean here w.r.t moving it into the type checker?

Re: diagnostics, I think it would be great addition to be able to emit diagnostics that say "*protocol* couldn't be synthesized because _such and such_ does not conform to *protocol*", but this is also something that we'd want across the board for `Codable` and future derived protocols as well. In the interest of scoping this PR, could that be something added to both later?